### PR TITLE
Surface_mesh:  Add switch for enabling/disabling Recycling of Garbage

### PR DESCRIPTION
**Pull request by afabri**
##  Summary of Changes

When calling a function such as `Surface_mesh::add_vertex()` one first reuses elements marked as removed.  We add a function to switch this off in case the user wants new elements.

`void set_recycle_garbage(bool b);`
_when set to true functions like add_vertex() will first see if they can recycle a previously element marked as removed. When set to false new elements will be created._

`bool does_recycle_garbage()`
_Getter_

## Release Management

* Affected package(s): Surface_mesh
* Feature/Small Feature (if any): [link](https://cgal.geometryfactory.com/CGAL/Members/wiki/Features/Small_Features/Surface_mesh-enable_recycling) -- **Pre-approval date:** 2020/05/22
* License and copyright ownership: GeometryFactory

### CODE REVIEW

File: Surface_mesh/include/CGAL/Surface_mesh/Surface_mesh.h [hunk 1]
```diff
@@ -1337,10 +1338,17 @@ class Surface_mesh
...
     template <typename Visitor>
     void collect_garbage(Visitor& visitor);
 
+    /// when set to `true` functions like `add_vertex()` will first
+    /// see if they can recycle a previously element marked as removed.
+    /// When set to `false` new elements will be created.
+    void set_recycle_garbage(bool b);
+
+    /// Getter
+    bool does_recycle_garbage();
```
Comment (lrineau) [inline Surface_mesh/include/CGAL/Surface_mesh/Surface_mesh.h:1351]:
```suggestion
    bool does_recycle_garbage() const;
```

File: Surface_mesh/include/CGAL/Surface_mesh/Surface_mesh.h [hunk 2]
```diff
@@ -3159,6 +3169,23 @@ collect_garbage()
...
+set_recycle_garbage(bool b)
+{
+  recycle_ = b;
+}
+
+template <typename P>
+bool
+Surface_mesh<P>::
+does_recycle_garbage()
+{
```
Comment (lrineau) [inline Surface_mesh/include/CGAL/Surface_mesh/Surface_mesh.h:3185]:
```suggestion
template <typename P>
bool
Surface_mesh<P>::
does_recycle_garbage() const
{
```

**Comment by sloriot** [comment]
You should set up an emacs hook for trailing whitespaces. See [here](https://www.emacswiki.org/emacs/DeletingWhitespace#toc3).

**Comment by afabri** [comment]
@sloriot I just added the small feature.

File: Surface_mesh/include/CGAL/Surface_mesh/Surface_mesh.h [hunk 3]
```diff
@@ -1337,10 +1338,17 @@ class Surface_mesh
...
     void collect_garbage();
 
-    //undocumented convenience fucntion that allows to get old-index->new-index information
+    //undocumented convenience function that allows to get old-index->new-index information
     template <typename Visitor>
     void collect_garbage(Visitor& visitor);
 
+    /// when set to `true` functions like `add_vertex()` will first
+    /// see if they can recycle a previously element marked as removed.
+    /// When set to `false` new elements will be created.
```
Comment (sloriot) [inline Surface_mesh/include/CGAL/Surface_mesh/Surface_mesh.h:1347]:
```suggestion
    /// controls the recycling or not of simplices previously marked as removed
    /// upon addition of new elements.
    /// When set to `true` (default value), new elements are first picked in the garbage (if any) 
    /// while if set to `false` only new elements are created.
```

**Comment by sloriot** [comment]
```
The following files have trailing whitespaces:

Surface_mesh/include/CGAL/Surface_mesh/Surface_mesh.h
```

**Comment by afabri** [comment]
@lrineau to me it seems ready to be tested, but I don't simply want to change the label.

**Comment by lrineau** [comment]
> @lrineau to me it seems ready to be tested, but I don't simply want to change the label.

I have modified the label.

As we are preparing CGAL-5.1-beta1, and this PR is for 5.2, I do not think it will be tested this week.

**Comment by lrineau** [comment]
@sloriot You set the `TODO` label. Was it for the feature page or for something else?

**Comment by sloriot** [comment]
yes

**Comment by maxGimeno** [comment]
Runtime error : https://cgal.geometryfactory.com/CGAL/testsuite/results-5.1-Ic-199.shtml#Surface_mesh

**Comment by afabri** [comment]
> 
> 
> Runtime error : https://cgal.geometryfactory.com/CGAL/testsuite/results-5.1-Ic-199.shtml#Surface_mesh

@maxGimeno can you please run it under a debugger as for VC++/Windows it is fine.  Thanks.

**Comment by maxGimeno** [comment]
The values returned by num_vertices, number_of_vertices and number_of_removed_vertices() are not the ones that you expect on some platforms. On my Ubuntu for example, .after the 10 removals, num_vertices() and number_of_removed_vertices() returns 10 and number_of_vertices() returns 0.

**Comment by maxGimeno** [comment]
Actually you never initialized _recycle, that must be why.

**Comment by maxGimeno** [comment]
Successfully tested in https://cgal.geometryfactory.com/CGAL/testsuite/results-5.1-Ic-200.shtml

**Comment by maxGimeno** [comment]
There is a conflict in CHANGES.md

**Comment by lrineau** [comment]
@maxGimeno I have refreshed the PR. It is ready to be tested again.

**Comment by maxGimeno** [comment]
Successfully tested in https://cgal.geometryfactory.com/CGAL/testsuite/results-5.2-Ic-30.shtml.

**Pull request merged**